text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+  The renaming rules allow these addons to adopt V2 format without breaking their public API. New addons _should not_ use renaming rules because it's confusing when the imports people type don't align with their real dependencies.
+
+## Externals
+
+The `externals` property in **Ember package metadata** allows a V2 addon to declare specific imported modules that should not be resolved at build time. Instead they will be resolved at runtime using the traditional `loader.js` `require()`.
+
+When Embroider compiles V1 packages to V2 it does automatic externals detection.
+
+When publishing a native V2 package, any externals need to be listed explicitly in **Ember package metadata**.
+
+An example of when you may need `externals` is when you need to consume a script (not a module) that contains arbitrary `define()` statements. The modules defined by those statements aren't resolvable, in general, at build time. So attempts to import them will generate build errors. By listing them in externals, you can defer them until runtime where they will work.
+
+## Assets
+
+To provide **Assets**, a package includes the `public-assets` key in **Ember package metadata**. It's a mapping from local paths to app-relative URLs that should be available in the final app. For example:
+
+    "name": "my-addon",
+    "ember-addon": {
+      "version": 2,
+      "public-assets": {
+        "./public/image.png": "/my-addon/image.png"
+      }
+    }
+
+with:
+
+    my-addon
+    └── public
+        └── image.png
+
+will result in final build output:
+
+    dist
+    └── my-addon
+        └── image.png
+
+Notice that we’re _not_ choosing to include assets via explicit ECMA `import`. The reason is that fine-grained inclusion of asset files is not critical to runtime performance. Any assets that your app doesn’t actually need, it should never fetch. Assets are always things with their own URLs.
+
+If two V2 packages try to emit assets with the same public URL, that's a build error.
+
+> Q: Should we just automatically namespace them instead?
+> A: That was considered, but it makes backward compatibility harder, and public URLs are not always free to choose/change.
+
+## Build Hooks
+
+In today’s v1 addon packages, the `index.js` file is the main entrypoint that allows an addon to integrate itself with the overall ember-cli build pipeline. The same idea carries forward to v2, with some changes.
+
+It is no longer the `main` entrypoint of the package (see **Own Javascript**). Instead, it’s located via the `build` key in **Ember package metadata**, which should point at a Javascript file. `build` is optional — if you don’t have anything to say, you don’t need the file.
+
+It is now an ECMA module, not a CJS file. The default export is a class that implements your build hooks (there is no required base class).
+
+Here is a list of build hooks, each of which will have its own section below.
+
+- configure
+- configureDependencies
+- skipBabel
```

Comment:
With the above hooks only, addons that want to plugin into the build process, to process files owned by the *app*, so obviously not available at publish-time (of the addon), seem to not be possible in v2 format (yet). Things like `ember-cli-sass` for example. These use cases are currently supported by hooks like [`postprocessTree`](https://ember-cli.com/api/classes/Addon.html#method_postprocessTree) or [`setupPreprocessorRegistry`](https://ember-cli.com/api/classes/Addon.html#method_setupPreprocessorRegistry). 

Maybe the RFC should state more clearly the way forward here:
* are the three hooks mentioned above the only ones supported for v2 addons (for now), so with this RFC not all v1 addons are migratable to v2 (basically only addons that can implement their features at publish-time, like addons exposing new run-time parts (e.g. components) to the consumer
* which will be extended further by following RFCs? (this seems what the "Supporting References" chapter suggests, but this was easy to miss)
* or are the other v1 hooks still supported in a v2 addon for now, something which the original [SPEC draft](https://github.com/embroider-build/embroider/blob/master/SPEC.md) goes into more detail.

Reply:
I believe this is addressed in the [Scope of this RFC](https://github.com/emberjs/rfcs/blob/embroider/text/0000-embroider-v2-package-format.md#scope-of-this-rfc) section:

> No provision for:
> ...
> * preprocessing your parent package's code
> * modifying your parent package's babel config
> ...
>
> It is understood that all of these are legitimate things for Ember addons to do. Defining these capabilities within v2 packages will be done in followup RFCs. It is simply too much scope to cover in one RFC.

The three hooks are the only APIs currently in v2 addons, which means that some addons will not be able to migrate to the v2 format yet. The v1 format will continue to be supported however, it just means that those addons will not be able to benefit from things like tree-shaking, etc.

Reply:
Thanks for the clarification! So this basically confirms my suspicion here:

> which will be extended further by following RFCs? (this seems what the "Supporting References" chapter suggests, but this was easy to miss)

But given the already quite long text, as I said this was easy to miss. So for future readers maybe a little note here would make sense.